pydantic/_internal/_core_metadata.py

```diff
@@ -1,84 +1,105 @@
+    pydantic_js_annotation_functions: list[GetJsonSchemaFunction] | None = None,
+    pydantic_js_prefer_positional_arguments: bool | None = None,
+    pydantic_js_input_core_schema: CoreSchema | None = None,
+    pydantic_js_updates: JsonDict | None = None,
+    pydantic_js_extra: JsonDict | JsonSchemaExtraCallable | None = None,
+) -> None:
+    from ..json_schema import PydanticJsonSchemaWarning
+
+    """Update CoreMetadata instance in place. When we make modifications in this function, they
+    take effect on the `core_metadata` reference passed in as the first (and only) positional argument.
+
+    First, cast to `CoreMetadata`, then finish with a cast to `dict[str, Any]` for core schema compatibility.
+    We do this here, instead of before / after each call to this function so that this typing hack
+    can be easily removed if/when we move `CoreMetadata` to `pydantic-core`.
+
+    For parameter descriptions, see `CoreMetadata` above.
+    """
+    core_metadata = cast(CoreMetadata, core_metadata)
+
+    if pydantic_js_functions:
+        core_metadata.setdefault('pydantic_js_functions', []).extend(pydantic_js_functions)
+
+    if pydantic_js_annotation_functions:
+        core_metadata.setdefault('pydantic_js_annotation_functions', []).extend(pydantic_js_annotation_functions)
+
+    if pydantic_js_prefer_positional_arguments:
+        core_metadata['pydantic_js_prefer_positional_arguments'] = pydantic_js_prefer_positional_arguments
+
+    if pydantic_js_input_core_schema:
+        core_metadata['pydantic_js_input_core_schema'] = pydantic_js_input_core_schema
+
+    if pydantic_js_updates:
+        if (existing_updates := core_metadata.get('pydantic_js_updates')) is not None:
+            core_metadata['pydantic_js_updates'] = {**existing_updates, **pydantic_js_updates}
+        else:
+            core_metadata['pydantic_js_updates'] = pydantic_js_updates
+
+    if pydantic_js_extra is not None:
+        existing_pydantic_js_extra = core_metadata.get('pydantic_js_extra')
+        if existing_pydantic_js_extra is None:
+            core_metadata['pydantic_js_extra'] = pydantic_js_extra
+        if isinstance(existing_pydantic_js_extra, dict):
+            if isinstance(pydantic_js_extra, dict):
+                core_metadata['pydantic_js_extra'] = {**existing_pydantic_js_extra, **pydantic_js_extra}
+            if isinstance(pydantic_js_extra, Callable):
+                warn(
+                    'Composing `dict` and `callable` type `json_schema_extra` is not supported.'
+                    "If you'd like support for this behavior, please open an issue on pydantic.",
+                    PydanticJsonSchemaWarning,
+                )
+        if isinstance(existing_pydantic_js_extra, Callable):
+            # if ever there's a case of a callable, we'll just keep the last json schema extra spec
+            core_metadata['pydantic_js_extra'] = pydantic_js_extra
+
+    core_metadata = cast(Dict[str, Any], core_metadata)
```

Comment:
Actually, we do get some benefit here: without this, we get type errors one layer up if we don't cast before assigning `some_schema['metadata'] = core_metadata` because `metadata` is typed as `dict[str, Any]`

Reply:
`cast` only has effect in the current scope:

Code sample in [pyright playground](https://pyright-play.net/?enableExperimentalFeatures=true&code=LAKAZgTg9gtgBAFwJ4AcCWA7A5nNMVQQJwCCGSANHAMYCGAzglRAKYBuLtANgPrIotQoACYswcMAFcM1ABQIWjAFylyASjgBaAHxwAclAwsloOGcSLiAXhoMEs4WmoIA2owhV3AXSoLGaoRBzC0YXAHJaMK84GzCAIzDA4L9rW0YHJ1cySlUkHxCEAJBAlJVsmLgAbwBfQNYObj5UFnlLDTgAYlzQKRlW-1B6zl5%2BFpT2ruzQIA)

```python

from typing import Any, cast, reveal_type

def func(test: Any) -> None:
    test = cast(dict[str, str], test)

    test['a'] = 'b'

    test = cast(dict[Any, Any], test)


test: Any = {}

reveal_type(test)  # Any
func(test)
reveal_type(test)  # Any
```

Reply:
Hmm, I'm seeing issues in my IDE when I remove the casting in that function (I guess, really the issue is, the type hint on return is `dict[str, Any]`, and if I don't cast the return value, then we see an issue.

Reply:
Hum but we are not returning in this function?

Reply:
Oh yes you're right, not sure how I overlooked that. Let's hop on a call to sort out the rest of the typing issues.

Reply:
You're right here, not sure why I got confused. I've removed the last `cast`.